File: grpc_api/server/src/resource/set.rs
use tonic::{Request, Response, Status};
use uuid::Uuid;
use bbthings_database::Resource;
use bbthings_grpc_proto::resource::set::set_service_server::SetService;
use bbthings_grpc_proto::resource::set::{
    SetSchema, SetId, SetIds, SetName, SetOption, SetUpdate, SetMemberRequest, SetMemberSwap,
    SetTemplateSchema, SetTemplateId, SetTemplateIds, SetTemplateName, SetTemplateOption, 
    SetTemplateUpdate, SetTemplateMemberRequest, SetTemplateMemberSwap,
    SetReadResponse, SetListResponse, SetCreateResponse, SetChangeResponse, 
    TemplateReadResponse, TemplateListResponse, TemplateCreateResponse, TemplateChangeResponse
};
use crate::common::validator::{AccessValidator, AccessSchema};
use crate::common::utility::handle_error;

const READ_SET: &str = "read_set";
const CREATE_SET: &str = "create_set";
const UPDATE_SET: &str = "update_set";
const DELETE_SET: &str = "delete_set";
const CHANGE_SET_MEMBER: &str = "change_set_member";

pub struct SetServer {
    resource_db: Resource,
    token_key: Vec<u8>,
    accesses: Vec<AccessSchema>
}

impl SetServer {
    pub fn new(resource_db: Resource) -> Self {
        Self {
            resource_db,
            token_key: Vec::new(),
            accesses: Vec::new()
        }
    }
    pub fn new_with_validator(resource_db: Resource, token_key: &[u8], accesses: &[AccessSchema]) -> Self {
        const PROCEDURES: &[&str] = &[
            READ_SET, CREATE_SET, UPDATE_SET, DELETE_SET, CHANGE_SET_MEMBER
        ];
        Self {
            resource_db,
            token_key: token_key.to_vec(),
            accesses: Self::construct_accesses(accesses, PROCEDURES)
        }
    }
}

#[tonic::async_trait]
impl SetService for SetServer {

    async fn read_set(&self, request: Request<SetId>)
        -> Result<Response<SetReadResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.read_set(Uuid::from_slice(&request.id).unwrap_or_default()).await;
        let result = match result {
            Ok(value) => Some(value.into()),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetReadResponse { result }))
    }

    async fn list_set_by_ids(&self, request: Request<SetIds>)
        -> Result<Response<SetListResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.list_set_by_ids(
            request.ids.into_iter().map(|id| Uuid::from_slice(&id).unwrap_or_default()).collect::<Vec<Uuid>>().as_slice()
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetListResponse { results }))
    }

    async fn list_set_by_template(&self, request: Request<SetTemplateId>)
        -> Result<Response<SetListResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.list_set_by_template(Uuid::from_slice(&request.id).unwrap_or_default()).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetListResponse { results }))
    }

    async fn list_set_by_name(&self, request: Request<SetName>)
        -> Result<Response<SetListResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.list_set_by_name(&request.name).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetListResponse { results }))
    }

    async fn list_set_option(&self, request: Request<SetOption>)
        -> Result<Response<SetListResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.list_set_option(
            request.template_id.map(|id| Uuid::from_slice(&id).unwrap_or_default()),
            request.name.as_deref()
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetListResponse { results }))
    }

    async fn create_set(&self, request: Request<SetSchema>)
        -> Result<Response<SetCreateResponse>, Status>
    {
        self.validate(request.extensions(), CREATE_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.create_set(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            Uuid::from_slice(&request.template_id).unwrap_or_default(),
            &request.name,
            Some(&request.description)
        ).await;
        let id = match result {
            Ok(value) => value,
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetCreateResponse { id: id.as_bytes().to_vec() }))
    }

    async fn update_set(&self, request: Request<SetUpdate>)
        -> Result<Response<SetChangeResponse>, Status>
    {
        self.validate(request.extensions(), UPDATE_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.update_set(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            request.template_id.map(|x| Uuid::from_slice(&x).unwrap_or_default()),
            request.name.as_deref(),
            request.description.as_deref()
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetChangeResponse { }))
    }

    async fn delete_set(&self, request: Request<SetId>)
        -> Result<Response<SetChangeResponse>, Status>
    {
        self.validate(request.extensions(), DELETE_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.delete_set(Uuid::from_slice(&request.id).unwrap_or_default()).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetChangeResponse { }))
    }

    async fn add_set_member(&self, request: Request<SetMemberRequest>)
        -> Result<Response<SetChangeResponse>, Status>
    {
        self.validate(request.extensions(), CHANGE_SET_MEMBER)?;
        let request = request.into_inner();
        let result = self.resource_db.add_set_member(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            Uuid::from_slice(&request.device_id).unwrap_or_default(),
            Uuid::from_slice(&request.model_id).unwrap_or_default(),
            &request.data_index
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetChangeResponse { }))
    }

    async fn remove_set_member(&self, request: Request<SetMemberRequest>)
        -> Result<Response<SetChangeResponse>, Status>
    {
        self.validate(request.extensions(), CHANGE_SET_MEMBER)?;
        let request = request.into_inner();
        let result = self.resource_db.remove_set_member(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            Uuid::from_slice(&request.device_id).unwrap_or_default(),
            Uuid::from_slice(&request.model_id).unwrap_or_default()
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetChangeResponse { }))
    }

    async fn swap_set_member(&self, request: Request<SetMemberSwap>)
        -> Result<Response<SetChangeResponse>, Status>
    {
        self.validate(request.extensions(), CHANGE_SET_MEMBER)?;
        let request = request.into_inner();
        let result = self.resource_db.swap_set_member(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            Uuid::from_slice(&request.device_id_1).unwrap_or_default(),
            Uuid::from_slice(&request.model_id_1).unwrap_or_default(),
            Uuid::from_slice(&request.device_id_2).unwrap_or_default(),
            Uuid::from_slice(&request.model_id_2).unwrap_or_default()
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(SetChangeResponse { }))
    }

    async fn read_set_template(&self, request: Request<SetTemplateId>)
        -> Result<Response<TemplateReadResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.read_set_template(Uuid::from_slice(&request.id).unwrap_or_default()).await;
        let result = match result {
            Ok(value) => Some(value.into()),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateReadResponse { result }))
    }

    async fn list_set_template_by_ids(&self, request: Request<SetTemplateIds>)
        -> Result<Response<TemplateListResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.list_set_template_by_ids(
            request.ids.into_iter().map(|id| Uuid::from_slice(&id).unwrap_or_default()).collect::<Vec<Uuid>>().as_slice()
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateListResponse { results }))
    }

    async fn list_set_template_by_name(&self, request: Request<SetTemplateName>)
        -> Result<Response<TemplateListResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.list_set_template_by_name(&request.name).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateListResponse { results }))
    }

    async fn list_set_template_option(&self, request: Request<SetTemplateOption>)
        -> Result<Response<TemplateListResponse>, Status>
    {
        self.validate(request.extensions(), READ_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.list_set_template_option(
            request.name.as_deref()
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateListResponse { results }))
    }

    async fn create_set_template(&self, request: Request<SetTemplateSchema>)
        -> Result<Response<TemplateCreateResponse>, Status>
    {
        self.validate(request.extensions(), CREATE_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.create_set_template(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            &request.name,
            Some(&request.description)
        ).await;
        let id = match result {
            Ok(value) => value,
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateCreateResponse { id: id.as_bytes().to_vec() }))
    }

    async fn update_set_template(&self, request: Request<SetTemplateUpdate>)
        -> Result<Response<TemplateChangeResponse>, Status>
    {
        self.validate(request.extensions(), UPDATE_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.update_set_template(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            request.name.as_deref(),
            request.description.as_deref()
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateChangeResponse { }))
    }

    async fn delete_set_template(&self, request: Request<SetTemplateId>)
        -> Result<Response<TemplateChangeResponse>, Status>
    {
        self.validate(request.extensions(), DELETE_SET)?;
        let request = request.into_inner();
        let result = self.resource_db.delete_set_template(Uuid::from_slice(&request.id).unwrap_or_default()).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateChangeResponse { }))
    }

    async fn add_set_template_member(&self, request: Request<SetTemplateMemberRequest>)
        -> Result<Response<TemplateChangeResponse>, Status>
    {
        self.validate(request.extensions(), CHANGE_SET_MEMBER)?;
        let request = request.into_inner();
        let result = self.resource_db.add_set_template_member(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            Uuid::from_slice(&request.type_id).unwrap_or_default(),
            Uuid::from_slice(&request.model_id).unwrap_or_default(),
            &request.data_index
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateChangeResponse { }))
    }

    async fn remove_set_template_member(&self, request: Request<SetTemplateMemberRequest>)
        -> Result<Response<TemplateChangeResponse>, Status>
    {
        self.validate(request.extensions(), CHANGE_SET_MEMBER)?;
        let request = request.into_inner();
        let result = self.resource_db.remove_set_template_member(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            request.template_index as usize
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateChangeResponse { }))
    }

    async fn swap_set_template_member(&self, request: Request<SetTemplateMemberSwap>)
        -> Result<Response<TemplateChangeResponse>, Status>
    {
        self.validate(request.extensions(), CHANGE_SET_MEMBER)?;
        let request = request.into_inner();
        let result = self.resource_db.swap_set_template_member(
            Uuid::from_slice(&request.id).unwrap_or_default(),
            request.template_index_1 as usize,
            request.template_index_2 as usize
        ).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TemplateChangeResponse { }))
    }

}

impl AccessValidator for SetServer {

    fn token_key(&self) -> Vec<u8> {
        self.token_key.clone()
    }

    fn accesses(&self) -> Vec<AccessSchema> {
        self.accesses.clone()
    }

}

File: grpc_api/client_python/src/bbthings_grpc/resource/set.py
from ..proto.resource import set_pb2, set_pb2_grpc
from typing import Optional, List
from uuid import UUID
import grpc
from ._schema import SetSchema, SetTemplateSchema


def read_set(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetId(id=id.bytes)
        response = stub.ReadSet(request=request, metadata=resource.metadata)
        return SetSchema.from_response(response.result)

def list_set_by_ids(resource, ids: List[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetIds(ids=list(map(lambda x: x.bytes, ids)))
        response = stub.ListSetByIds(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def list_set_by_template(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateId(id=id.bytes)
        response = stub.ListSetByTemplate(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def list_set_by_name(resource, name: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetName(name=name)
        response = stub.ListSetByName(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def list_set_option(resource, template_id: Optional[UUID], name: Optional[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        template_bytes = None
        if template_id != None: template_bytes = template_id.bytes
        request = set_pb2.SetOption(
            template_id=template_bytes,
            name=name
        )
        response = stub.ListSetOption(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def create_set(resource, id: UUID, template_id: UUID, name: str, description: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetSchema(
            id=id.bytes,
            template_id=template_id.bytes,
            name=name,
            description=description
        )
        response = stub.CreateSet(request=request, metadata=resource.metadata)
        return UUID(bytes=response.id)

def update_set(resource, id: UUID, template_id: Optional[UUID], name: Optional[str], description: Optional[str]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        template_bytes = None
        if template_id != None: template_bytes = template_id.bytes
        request = set_pb2.SetUpdate(
            id=id.bytes,
            template_id=template_bytes,
            name=name,
            description=description
        )
        stub.UpdateSet(request=request, metadata=resource.metadata)

def delete_set(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetId(id=id.bytes)
        stub.DeleteSet(request=request, metadata=resource.metadata)

def add_set_member(resource, id: UUID, device_id: UUID, model_id: UUID, data_index: List[int]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetMemberRequest(
            id=id.bytes,
            device_id=device_id.bytes,
            model_id=model_id.bytes,
            data_index=bytes(data_index)
        )
        stub.AddSetMember(request=request, metadata=resource.metadata)

def remove_set_member(resource, id: UUID, device_id: UUID, model_id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetMemberRequest(
            id=id.bytes,
            device_id=device_id.bytes,
            model_id=model_id.bytes
        )
        stub.RemoveSetMember(request=request, metadata=resource.metadata)

def swap_set_member(resource, id: UUID, device_id_1: UUID, model_id_1: UUID, device_id_2: UUID, model_id_2: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetMemberSwap(
            id=id.bytes,
            device_id_1=device_id_1.bytes,
            model_id_1=model_id_1.bytes,
            device_id_2=device_id_2.bytes,
            model_id_2=model_id_2.bytes
        )
        stub.SwapSetMember(request=request, metadata=resource.metadata)

def read_set_template(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateId(id=id.bytes)
        response = stub.ReadSetTemplate(request=request, metadata=resource.metadata)
        return SetTemplateSchema.from_response(response.result)

def list_set_template_by_ids(resource, ids: List[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateIds(ids=list(map(lambda x: x.bytes, ids)))
        response = stub.ListSetTemplateByIds(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetTemplateSchema.from_response(result))
        return ls

def list_set_template_by_name(resource, name: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateName(name=name)
        response = stub.ListSetTemplateByName(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetTemplateSchema.from_response(result))
        return ls

def list_set_template_option(resource, name: Optional[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetOption(
            name=name
        )
        response = stub.ListSetTemplateOption(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def create_set_template(resource, id: UUID, name: str, description: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateSchema(
            id=id.bytes,
            name=name,
            description=description
        )
        response = stub.CreateSetTemplate(request=request, metadata=resource.metadata)
        return UUID(bytes=response.id)

def update_set_template(resource, id: UUID, name: Optional[str], description: Optional[str]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateUpdate(
            id=id.bytes,
            name=name,
            description=description
        )
        stub.UpdateSetTemplate(request=request, metadata=resource.metadata)

def delete_set_template(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateId(id=id.bytes)
        stub.DeleteSetTemplate(request=request, metadata=resource.metadata)

def add_set_template_member(resource, id: UUID, type_id: UUID, model_id: UUID, data_index: List[int]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateMemberRequest(
            id=id.bytes,
            type_id=type_id.bytes,
            model_id=model_id.bytes,
            data_index=bytes(data_index)
        )
        stub.AddSetTemplateMember(request=request, metadata=resource.metadata)

def remove_set_template_member(resource, id: UUID, index: int):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateMemberRequest(
            id=id.bytes,
            template_index=index
        )
        stub.RemoveSetTemplateMember(request=request, metadata=resource.metadata)

def swap_set_template_member(resource, id: UUID, index_1: int, index_2: int):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateMemberSwap(
            id=id.bytes,
            template_index_1=index_1,
            template_index_2=index_2
        )
        stub.SwapSetTemplateMember(request=request, metadata=resource.metadata)

